File: STL2AOP_Deno/output.ts
// Import necessary libraries in Deno
import { PDFDocument, rgb } from "https://deno.land/x/pdf@1.4.0/mod.ts";
import { writeCSV } from "https://deno.land/x/csv/mod.ts";
import { ensureDirSync } from "https://deno.land/std/fs/mod.ts";
import { join } from "https://deno.land/std/path/mod.ts";

// Helper function to get downloads folder in Deno (cross-platform)
function getDownloadsFolder(): string {
	const home = Deno.env.get("HOME") || Deno.env.get("USERPROFILE");
	const downloadsPath = join(home!, "Downloads");
	ensureDirSync(downloadsPath); // Create downloads folder if it doesn't exist
	return downloadsPath;
}

// Function to generate a PDF file using pdf-lib
async function generatePDF(
	lngths: number[],
	perimeters: number[],
	CSA: number[],
	APW: number[],
	MLW: number[],
): Promise<string> {
	const downloadsPath = getDownloadsFolder();
	const outputPath = join(downloadsPath, "ampscanReport.pdf");

	// Load existing PDF templates
	const templateBytes = await Deno.readFile("Measurements Template.pdf");
	const outputTemplateBytes = await Deno.readFile("Output Template.pdf");

	const pdfDoc = await PDFDocument.load(templateBytes);
	const outputDoc = await PDFDocument.load(outputTemplateBytes);

	// Create a new page or modify the existing page
	const [templatePage] = await pdfDoc.copyPages(outputDoc, [0]);

	const page = pdfDoc.getPage(0);
	const font = await pdfDoc.embedFont("Courier-Bold");
	page.setFont(font);

	// Add custom drawings
	for (let i = 1; i < lngths.length - 1; i++) {
		const stringl = `${Math.abs(Math.round(lngths[i] * 10) / 10)}`;
		const stringp = `${Math.abs(Math.round(perimeters[i] * 10) / 10)}`;
		page.drawText(stringl, { x: 360 + (i - 1) * 27, y: 474 - (i - 1) * 41.5 });
		page.drawText(stringp, { x: 88, y: 524.5 - (i - 1) * 74.5 });
	}

	const stringMaxZ = `${Math.abs(Math.round(lngths[lngths.length - 1] * 10) / 10)}`;
	page.drawText(stringMaxZ, { x: 514, y: 419 });

	// Add images (you may need a compatible image processing method)
	const antImage = await pdfDoc.embedPng(await Deno.readFile("ant.png"));
	const latImage = await pdfDoc.embedPng(await Deno.readFile("lat.png"));
	const figureImage = await pdfDoc.embedPng(await Deno.readFile("figure.png"));

	page.drawImage(antImage, { x: 38, y: 225, width: 256, height: 256 });
	page.drawImage(latImage, { x: 300, y: 225, width: 256, height: 256 });
	page.drawImage(figureImage, { x: -2.5, y: -50, width: 334, height: 200 });

	for (let i = 1; i < CSA.length; i += 2) {
		const sCSA = `${Math.round(CSA[i] * 10) / 10}`;
		const sAPW = `${Math.round(APW[i] * 10) / 10}`;
		const sMLW = `${Math.round(MLW[i] * 10) / 10}`;
		page.drawText(sCSA, { x: 403, y: 145 - (i - 1) * 11.5 });
		page.drawText(sAPW, { x: 465, y: 145 - (i - 1) * 11.5 });
		page.drawText(sMLW, { x: 520, y: 145 - (i - 1) * 11.5 });
	}

	// Save the updated PDF
	const pdfBytes = await pdfDoc.save();
	await Deno.writeFile(outputPath, pdfBytes);

	return outputPath;
}

// Function to generate a CSV for regular bins
async function generateRegBinsCsv(
	file: string,
	regObject: any,
	numBins: number,
	scalarMin: number,
	scalarMax: number,
) {
	const binSize = (scalarMax - scalarMin) / numBins;
	const bins = new Array(numBins).fill(0);
	const binValues = Array.from(
		{ length: numBins },
		(_, i) => scalarMin + binSize * i,
	);

	regObject.values.forEach((point: number) => {
		const bin = Math.floor((point - scalarMin) / binSize);
		if (bin < 0) {
			bins[0]++;
		} else if (bin >= bins.length) {
			bins[bins.length - 1]++;
		} else {
			bins[bin]++;
		}
	});

	const totalValues = regObject.values.length;
	const csvData = binValues.map((val, i) => [val, bins[i] / totalValues]);

	const outputFilePath = join(Deno.cwd(), file);
	await writeCSV(outputFilePath, csvData);
}

// Function to generate a simple CSV for regObject values
async function generateRegCsv(file: string, regObject: any) {
	const csvData = regObject.values.map((value: number) => [value]);

	const outputFilePath = join(Deno.cwd(), file);
	await writeCSV(outputFilePath, csvData);
}

// Function to generate specific output for a registration object and write it to a CSV
async function generateSpec(file: string, regObject: any) {
	const absMean = Math.abs(
		regObject.values.reduce((a: number, b: number) => a + b) /
			regObject.values.length,
	);
	const absStd = Math.sqrt(
		regObject.values
			.map((x: number) => Math.pow(Math.abs(x) - absMean, 2))
			.reduce((a: number, b: number) => a + b) / regObject.values.length,
	);
	const mean =
		regObject.values.reduce((a: number, b: number) => a + b) /
		regObject.values.length;
	const std = Math.sqrt(
		regObject.values
			.map((x: number) => Math.pow(x - mean, 2))
			.reduce((a: number, b: number) => a + b) / regObject.values.length,
	);
	const valuemin = Math.min(...regObject.values);
	const valuemax = Math.max(...regObject.values);

	const outDict = {
		"mean distance": mean,
		"standard deviation": std,
		"minimum distance": valuemin,
		"maximum distance": valuemax,
		"mean absolute distance": absMean,
		"absolute standard deviation": absStd,
		// Add more properties as needed
	};

	const csvData = Object.entries(outDict).map(([key, value]) => [key, value]);

	const outputFilePath = join(Deno.cwd(), file);
	await writeCSV(outputFilePath, csvData);
}

File: STL2AOP_Provel_Python/custom_amp1/output.py
from PyPDF2 import PdfFileReader, PdfFileWriter
import pathlib
from pathlib import Path
from reportlab.pdfgen import canvas
import io
import os
import csv
import numpy as np

def getPDF(lngths, perimeters, CSA, APW, MLW):
    """
    Creates a PDF file containing information about the limb in correct
    locations on the page, then merges the PDF file with the existing template
    to create the output file.

    Returns
    -------
    The file path to the PDF
    """
    path = pathlib.Path(__file__).parent.absolute()
    my_file = Path(path, "Measurements Template.pdf")
    try:
        my_abs_path = my_file.resolve(strict=True)
    except FileNotFoundError:
        return 1 

    packet = io.BytesIO()
    c = canvas.Canvas(packet)
    for i in range(1, len(lngths) - 1):
        stringl = "{}".format(abs(round(lngths[i], 1)))
        stringp = "{}".format(abs(round(perimeters[i], 1)))
        c.drawString(360 + ((i - 1) * 27), 474 - ((i - 1) * 41.5), stringl)
        c.drawString(88, 524.5 - ((i - 1) * 74.5), stringp)

    stringmaxZ = "{}".format(abs(round(lngths[len(lngths) - 1], 1)))
    c.drawString(514, 419, stringmaxZ)
    c.setFont("Courier-Bold", 12)
    c.drawString(65, 575, "Perimeter / cm")
    c.drawString(400, 520, "Distance / cm")
    c.showPage()
    c.drawImage("ant.png", 38, 225, 256, 256)
    c.drawImage("lat.png", 300, 225, 256, 256)
    c.drawImage("figure.png", -2.5, -50, 334, 200)
    for i in range(1, len(CSA), 2):
        sCSA = "{}".format(round(CSA[i], 1))
        sAPW = "{}".format(round(APW[i], 1))
        sMLW = "{}".format(round(MLW[i], 1))
        c.drawString(403, 145 - ((i - 1) * 11.5), sCSA)
        c.drawString(465, 145 - ((i - 1) * 11.5), sAPW)
        c.drawString(520, 145 - ((i - 1) * 11.5), sMLW)
    c.save()
    packet.seek(0)
    newpdf = PdfFileReader(packet)

    template = PdfFileReader(open(os.path.join(path, "Measurements Template.pdf"), "rb"))
    t2 = PdfFileReader(open(os.path.join(path, "Output Template.pdf"), "rb"))
    output = PdfFileWriter()
    page = t2.getPage(0)
    page.mergePage(newpdf.getPage(1))
    page2 = template.getPage(0)
    page2.mergePage(newpdf.getPage(0))
    output.addPage(page)
    output.addPage(page2)

    output_file_path = os.path.join(get_downloads_folder(), "ampscanReport.pdf")
    with open(output_file_path, "wb") as outputStream:
        output.write(outputStream)

    return output_file_path

def get_downloads_folder():
    """Gets the downloads folder in a relatively platform independent way"""
    downloads_path = os.path.join(os.path.expanduser("~"), "Downloads")
    if not os.path.exists(downloads_path):  # If downloads folder doesn't exist, create it
        os.mkdir(downloads_path)
    return downloads_path

def generateRegBinsCsv(file, regObject, numBins, scalarMin, scalarMax):
    """
    Generates a CSV file output of scalar values put into bins
    :param file: The open file to save csv output to. Should be open with newline=''
    :param regObject: The reg object with scalar values
    :param numBins: Number of bins for scalar values
    :param scalarMin: The min scalar value to look for
    :param scalarMax: The max scalar value to look for
    :return: None
    """
    writer = csv.writer(file)

    binSize = (scalarMax - scalarMin) / numBins
    bins = np.zeros(numBins, dtype=np.int32)
    binValues = np.linspace(scalarMin, scalarMax, numBins, endpoint=False)
    
    for point in regObject.values:
        bin = int((point - scalarMin) / binSize)
        if bin < 0:
            bins[0] += 1
        elif bin >= len(bins):
            bins[-1] += 1
        else:
            bins[bin] += 1

    total_values = len(regObject.values)

    for i in range(numBins):
        writer.writerow([scalarMin + binSize * i, bins[i] / total_values])

def generateRegCsv(file, regObject):
    """
    Generates a CSV file output of scalar values put into bins
    :param file: The open file to save csv output to. Should be open with newline=''
    :param regObject: The reg object with scalar values
    :return: None
    """
    writer = csv.writer(file)
    for value in regObject.values:
        writer.writerow([value])

def generate_spec(file, regObject):
    """
    This function automatically generates specific output for the specified
    registration object.
    
    Parameters
    ----------
    regObject: AmpObject
        The registration object.

    Returns
    -------
    None    
    """
    absmean = np.mean(np.abs(regObject.values), dtype=np.float64)
    absstd = np.std(np.abs(regObject.values), dtype=np.float64)
    mean = np.mean(regObject.values, dtype=np.float64)
    std = np.std(regObject.values, dtype=np.float64)
    valuemin = np.min(regObject.values, dtype=np.float64)
    valuemax = np.max(regObject.values, dtype=np.float64)

    idxleft = np.where(regObject.vert[:, 0] >= 0)
    idxright = np.where(regObject.vert[:, 0] <= 0)
    valueleft = regObject.values[idxleft]
    valueright = regObject.values[idxright]
    meanleft = np.mean(valueleft, dtype=np.float64)
    stdleft = np.std(valueleft, dtype=np.float64)
    meanright = np.mean(valueright, dtype=np.float64)
    stdright = np.std(valueright, dtype=np.float64)

    gap = np.where(regObject.values > 0)[0]
    integratedgap = np.sum(regObject.values[gap], dtype=np.float64) / regObject.values.shape[0]
    gap_percentage = (gap.shape[0]) / (regObject.values.shape[0]) * 100
    pressure = np.where(regObject.values < -3)[0]
    integratedHP = np.sum(regObject.values[pressure], dtype=np.float64) / regObject.values.shape[0]
    pressure_percentage = (pressure.shape[0]) / (regObject.values.shape[0]) * 100

    outdict = {
        'mean distance': mean,
        'standard deviation': std,
        'minimum distance': valuemin,
        'maximum distance': valuemax,
        'mean absolute distance': absmean,
        'absolute standard deviation': absstd,
        'Left mean distance': meanleft,
        'Left standard deviation': stdleft,
        'Right mean distance': meanright,
        'Right standard deviation': stdright,
        'percentage of gap area': gap_percentage,
        'percentage of high pressure area': pressure_percentage,
        'percentage of area within range': 100 - pressure_percentage - gap_percentage,
        'integrated value of gap area': integratedgap,
        'integrated value of high pressure area': integratedHP
    }

    with open(os.getcwd() + file, 'w', newline='') as myfile:
        writer = csv.DictWriter(myfile, fieldnames=['Name', 'Value'])
        for key, value in outdict.items():
            writer.writerow({'Name': key, 'Value': value})

